### PR TITLE
feat(utilities): add possibility to import single functions by appending them to the import path.

### LINKED ISSUE #452: request: ability to import single functions in `@sapphire/utilities`
### Is there an existing issue or pull request for this?

- [X] I have searched the existing issues and pull requests

### Feature description

Right now, if we want to import an utility, we have to import from `@sapphire/utilities`, which loads all of the utilities, which in turn increases the load time and memory usage.

### Desired solution

Be able to import with subpaths, e.g. `import { chunk } from '@sapphire/utilities/chunk';`

### Alternatives considered

N/a.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by imranbarbhuiya**
fixes #452

tsup was bunding the dts in one file and has no option to disable it. So I've used `tsc` to build the types files.

### CODE REVIEW

**Comment by imranbarbhuiya** [comment]
Not sure why build is failing with targe `2021`

**Review by favna — CHANGES_REQUESTED** [review]
Granted even CJS didn't work before my commit above this review, this PR still needs more work.

Create a file on the root of the repository called `test.cjs` with content:
```js
/* eslint-disable @typescript-eslint/no-var-requires */
const { isFunction } = require('@sapphire/utilities/isFunction');

console.log(isFunction(() => undefined));
```
then run it with `node test.cjs`. This will return `true` as expected.

However doing the same with `test.mjs` and content:
```js
import { isFunction } from '@sapphire/utilities/isFunction';

console.log(isFunction(() => undefined));
```

will give:
```
node file.mjs
file:///Users/favna/workspace/sapphire/utilities/file.mjs:1
import { isFunction } from '@sapphire/utilities/isFunction';
         ^^^^^^^^^^
SyntaxError: Named export 'isFunction' not found. The requested module '@sapphire/utilities/isFunction' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from '@sapphire/utilities/isFunction';
const { isFunction } = pkg;

    at ModuleJob._instantiate (node:internal/modules/esm/module_job:123:21)
    at async ModuleJob.run (node:internal/modules/esm/module_job:189:5)
    at async Promise.all (index 0)
    at async ESMLoader.import (node:internal/modules/esm/loader:528:24)
    at async loadESM (node:internal/process/esm_loader:91:5)
    at async handleMainPromise (node:internal/modules/run_main:65:12)

Node.js v18.8.0
```

And lastly, TS is also broken, again with a similar file this time `file.ts`:

```ts
import { isFunction } from '@sapphire/utilities/isFunction';

console.log(isFunction(() => undefined));
```

This gives a compiler error:
```
Cannot find module '@sapphire/utilities/isFunction' or its corresponding type declarations.ts(2307)
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/utilities/package.json [hunk 1]
```diff
@@ -32,7 +256,6 @@
 	},
 	"files": [
 		"dist/**/*.js*",
-		"dist/**/*.mjs*",
```
[the remark this answers is not in the document]
Comment (imranbarbhuiya) [reply packages/utilities/package.json:35]: I removed it when `legacy` option was true. Will revert it again

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/utilities/package.json [hunk 2]
```diff
@@ -10,20 +10,244 @@
...
+		".": {
+			"types": "./dist/index.d.ts",
+			"import": "./dist/index.mjs",
+			"require": "./dist/index.js",
+			"browser": "./dist/index.global.js"
+		},
+		"./arrayStrictEquals": {
+			"types": "./dist/lib/arrayStrictEquals.d.ts",
+			"import": "./dist/lib/arrayStrictEquals.mjs",
+			"require": "./dist/lib/arrayStrictEquals.js"
```
[the remark this answers is not in the document]
Comment (imranbarbhuiya) [reply packages/utilities/package.json:22]: Export maps are valid for node only.  So bundler won't check it (didn't tested tho).  Also tsup has no way to define global Name file wise so it'll make all the files with same name.  If I use roll up to trees hake those files,  roll up will add a global name based on file name.  So maybe we can use this but I'm not sure if export map will work or not

File: packages/discord.js-utilities/src/lib/PaginatedMessages/PaginatedMessage.ts [hunk 3]
```diff
@@ -538,7 +541,7 @@ export class PaginatedMessage {
 	 * ```
 	 */
 	public addAsyncPageBuilder(builder: MessageBuilder | ((builder: MessageBuilder) => Promise<MessageBuilder>)): this {
-		return this.addPage(async () => (isFunction(builder) ? builder(new MessageBuilder()) : builder));
```
[the remark this answers is not in the document]
Comment (favna) [reply packages/discord.js-utilities/src/lib/PaginatedMessages/PaginatedMessage.ts:541]: I don't think the async added anything? Without anything being awaited and `builder` already returning a ... oh... oh wait.. `builder(...)` returned a Promise` but just `builder` doesn't without async. Yeah this should be reverted.

**Review by vladfrangu — APPROVED** [review]

**Review by favna — CHANGES_REQUESTED** [review]
Can you update this to use node-utilities now @imranbarbhuiya ?

**Review by favna — APPROVED** [review]

File: scripts/aliases/utilities.mjs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+export const aliasStore = new Map([
+	['filterNullAndUndefined', 'filterNullish'],
+	['filterNullAndUndefinedAndEmpty', 'filterNullishOrEmpty'],
+	['filterNullAndUndefinedAndZero', 'filterNullishOrZero'],
+	['isNullOrUndefined', 'isNullish'],
+	['isNullOrUndefinedOrEmpty', 'isNullishOrEmpty'],
+	['isNullOrUndefinedOrZero', 'isNullishOrZero'],
+	['sleep', 'sleepSync']
```
Comment (kyranet) [inline scripts/aliases/utilities.mjs:8]: `sleep` and `sleepSync` aren't the same, they should be split.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: scripts/aliases/utilities.mjs (hunk 4 above)
Comment (favna) [reply scripts/aliases/utilities.mjs:8]: The aliases are to indicate they target the same file in the mapping, not that the functions are actual aliases.

**Review by kyranet — APPROVED** [review]

**Pull request merged**